### ISSUE TITLE
CA1508 mis-fires, alleges dead code incorrectly

### ISSUE THREAD

**Issue by AArnott** [dotnet/roslyn-analyzers#4552]
**Version Used**: VS 16.9 (30809.25.main)

**Steps to Reproduce**:

```cs
using System;

#nullable enable

public class C {
    private object? value = null;
    
    public string Value => (this.value != null) ? ((this.value as string) ?? "hi") : "null";    
}
```

**Expected Behavior**:

No warnings.

**Actual Behavior**:

CA1508 mis-fires, claiming:

> 'this.value as string' is never 'null'. Remove or refactor the condition(s) to avoid dead code.

But of course this is wrong, since `this.value as string` will be `null` if (for example) `this.value` is set to `5` (boxed).

![image](https://user-images.githubusercontent.com/3548/101695631-234da900-3a32-11eb-96b6-3767d317ff7d.png)

**Comment by mavasani** [comment]
Verified that the issue repros on analyzers from .NET 5.0 SDK as well as `Microsoft.CodeAnalysis.NetAnalyzers` NuGet package `5.0.0`. However, the issue is fixed in master and `Microsoft.CodeAnalysis.NetAnalyzers` NuGet package `5.0.1`. The newer package has not been inserted into .NET 5 SDK, but should be available in [daily builds](https://github.com/dotnet/installer#installers-and-binaries) of .NET 6 SDK or you can directly reference the latest NuGet package itself to get the analyzers with the fix.

Closing the issue.

**Comment by mavasani** [comment]
Note that because .NET 5 SDK itself has a much higher bug bar, all fixes to analyzers which are off by default (like CA1508) which were made _after the release_ of initial .NET 5.0 SDK would only be available in .NET 6 SDK or the NuGet package releases of `Microsoft.CodeAnalysis.NetAnalyzers` on nuget.org.
